messages/compiler/walker.py

```
@@ -0,0 +1,86 @@
+# Concord
+#
+# Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+from exceptions import CmfParseError
+
+
+def is_primitive(type):
+    """Return true if a type is a primitive"""
+    return type in [
+        "uint8", "uint16", "uint32", "uint64", "int8", "int16", "int32",
+        "int64", "bool", "string", "bytes"
+    ]
+
+
+def field_type(type):
+    """Extract the outermost type from a field"""
+    if not isinstance(type, dict):
+        if is_primitive(type):
+            return type
+        return "msg"
+    for compound in ["list", "kvpair", "map", "optional", "oneof"]:
+        if compound in type:
+            return compound
```

Looks like I need to throw an exception here if nothing is found.

---

fixed.